### PR TITLE
Chnage concurrency to :shared and publish outside of synchronized code

### LINKED ISSUE #59: Batches should be published in paralel when using multiple pipeline workers
Given that this plugin has `concurrency :single` (meaning a single instance will be shared by workers) the behavior of `Batcher.enqueue` causes the sending to be serialized across workers causing a bottleneck.

I believe the desired behavior is to have all workers sending data independently of each other whenever they have a batch ready.

### PR DESCRIPTION
**Pull request by HonzaKral**
Fixes #59

### CODE REVIEW

**Comment by josephlewis42** [comment]
I think this looks good. I've been out of Ruby development for about a year now so it would help for someone else to take a peek.

File: lib/logstash/outputs/bigquery/batcher.rb [hunk 1]
```diff
@@ -49,15 +50,14 @@ def enqueue(message)
...
-              yield(orig) if block_given?
-              return orig
             end
 
-            nil
           ensure
             @lock.write_lock.unlock
           end
+
+          yield(orig) if block_given? && !orig.nil?
```
Comment (jsvd) [inline lib/logstash/outputs/bigquery/batcher.rb:59]: Seeing how (1) it copies the batch to `orig`, (2) doesn't mutate the object in `@batch`, (3) executes `clear` from within the lock, I think we're safe in `enqueue`.

As for [`publish`](https://github.com/logstash-plugins/logstash-output-google_bigquery/blob/7822aa688474148c1c5059f611879c86cda29411/lib/logstash/outputs/google_bigquery.rb#L232), we don't rely on any instance state, so as long as the google java client we create in [`initialize_google_client`](https://github.com/logstash-plugins/logstash-output-google_bigquery/blob/7822aa688474148c1c5059f611879c86cda29411/lib/logstash/outputs/bigquery/streamclient.rb#L95-L107) is thread safe, then I think we're good to merge here.

**Review by josephlewis42 — APPROVED** [review]

**Comment by karenzone** [comment]
@HonzaKral  Thanks for making this change. It appears to be ready to merge.

**Comment by jsvd** [comment]
merged in 1196273

**Comment by jsvd** [comment]
this has been released as v4.1.4. thanks for the contribution @HonzaKral and review @josephlewis42!

**Pull request closed without merging**